predict_proba output column

prithvikannan opened the pull request:
Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## Related Issues/PRs

<!--
Please reference any related feature requests, issues, or PRs here. For example, `#123`. To automatically close the referenced issues when this PR is merged, please use a closing keyword (close, fix, or resolve). For example, `Resolve #123`. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.
-->

<!-- Resolve --> #xxx

## What changes are proposed in this pull request?

Worst predictions table
<img width="993" alt="image" src="https://user-images.githubusercontent.com/46332835/198693267-934f7a5c-3b91-474c-bb62-d9a1afe3bf80.png">

Predicted training data written to parquet
<img width="1093" alt="image" src="https://user-images.githubusercontent.com/46332835/198693422-20395944-2a2b-4642-8856-fbdd567a2507.png">



## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->
- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

prithvikannan committed:
predict_proba output column

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

mlflow-automation commented:
<!-- documentation preview -->

### Documentation preview will be available [here](https://output.circle-artifacts.com/output/job/cd5f40da-9201-42ea-90c5-477d2ee247e6/artifacts/0/docs/build/html/index.html).

<details>
<summary>Notes</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated on every commit to this PR.
- Job URL: https://circleci.com/gh/mlflow/mlflow/34673
- Updated at: 2022-10-27 23:38:48.337373

</details>

prithvikannan committed:
log loss scoring

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan commented:
@jinzhang21 i have added a couple screenshots to the PR description. 

in the worst predictions table - i have replaced the `predictions` column with the predicted probabilities. in the predicted training data (written to parquet), i have added a new column called `predicted_probability` in addition to `predictions`

also, when users want to use the model for inference, they will still need to call `model.predict_proba` themselves. 

does this UX accomplish the goal?

BTW, `pipeline.yaml` changes are WIP and not in this PR yet

prithvikannan committed:
gate on pipeline yaml

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan commented:
`pipeline.yaml` changes now added

bbarnes52 commented:
Excellent work!

PTAL at https://github.com/mlflow/mlflow/pull/6631 as well. I believe we may also want to thread through the configured value for `output_probabilities` to the `save_model` function [https://github.com/mlflow/mlflow/blob/master/mlflow/pipelines/steps/train.py#L320](here), e.g. if `output_probabilities = True` then `pyfunc_predict_fn="predict_proba"`

As far as worst examples go, I'd be inclined to always use `predict_proba` if it's available

Wdyt?

prithvikannan committed:
pyfunc predict proba

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan commented:
> Excellent work!
> 
> PTAL at #6631 as well. I believe we may also want to thread through the configured value for `output_probabilities` to the `save_model` function [https://github.com/mlflow/mlflow/blob/master/mlflow/pipelines/steps/train.py#L320](here), e.g. if `output_probabilities = True` then `pyfunc_predict_fn="predict_proba"`

Good point! I've updated the PR to include this behavior.

> As far as worst examples go, I'd be inclined to always use `predict_proba` if it's available
> 
> Wdyt?

IMO, this would create a confusing UX where users do not see classes (ie. setosa or versicolor) by default. My preference is to keep this as is @bbarnes52

dbczumar commented on mlflow/pipelines/steps/train.py:
```diff
@@ -317,7 +318,13 @@ def my_warn(*args, **kwargs):
                     shutil.rmtree(output_model_path)
-                mlflow.sklearn.save_model(model, output_model_path)
+
+                if self.output_probabilities:
+                    mlflow.sklearn.save_model(
+                        model, output_model_path, pyfunc_predict_fn="predict_proba"
+                    )
+                else:
+                    mlflow.sklearn.save_model(model, output_model_path)
```
Looks like we're logging the model to MLflow Tracking on lines 311 - 313 and then saving the model to local storage with a different pyfunc predict function here. Should we be extending lines 311 - 313 instead of adding a separate `save_model` call here?

We should also make sure that the model schema is adjusted based on whether or not we're outputting probabilities.

prithvikannan replied:
Thanks for calling this out. I've updated the call to `mlflow.sklearn.log_model` to use predict_proba and use the updated schema. I think that we need to also keep these changes in `mlflow.sklearn.save_model` if a user wants to load from local.

prithvikannan committed:
use predict proba for schema and log_model

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

bbarnes52 commented:
> > As far as worst examples go, I'd be inclined to always use `predict_proba` if it's available
> > Wdyt?
> 
> IMO, this would create a confusing UX where users do not see classes (ie. setosa or versicolor) by default. My preference is to keep this as is @bbarnes52

I did a poor job communicating my point, mea culpa. What I had in mind is that probabilities are used to rank the worst examples regardless of the configured value of `output_probabilities`. To your point, I agree the classes should also be included - definitely if `output_probabilities` is False and potentially also if it's True (no strong opinion here though).

jinzhang21 reviewed (commented):
The code change LGTM as is. However I have concerns on the product behavior as the following. Let's set up a meeting with training team to discuss what's the best UX in this case (looping in Harsh or Lu).
- Users often want to have both `predict` and `predict_proba`, with `predict` as the primary function to classify input and `predict_proba` as the secondary method to gain confidence in prediction power. It makes more sense to me to add `predict_proba` method to the Pyfunc model instead of replacing the `predict` method and its output under the hood.
- `predict_proba`'s output is a N-class vector, whereas `predict`'s output is a scala (usually). The step card showing probabilities becomes hard to understand in multi-class scenario if it's always the probabilities of the `positive_label` is shown. Instead, it might be more useful to show something like `predicted_class` column +  `probability_for_the_predicted_class` column.

jinzhang21 commented:
> > * Users often want to have both `predict` and `predict_proba`, with `predict` as the primary function to classify input and `predict_proba` as the secondary method to gain confidence in prediction power. It makes more sense to me to add `predict_proba` method to the Pyfunc model instead of replacing the `predict` method and its output under the hood.
> 
> Do we have any examples of having multiple pyfunc predict methods? In this PR #6631 we support changing the predict method, but I don't see a way to pass in both `predict` and `predict_proba` to pyfunc model.

You are right it might be difficult to support both w/o adding predict_proba to Pyfunc base class first. Let's merge the PR as is for now then.

jinzhang21 commented on mlflow/pipelines/steps/train.py:
```diff
@@ -159,6 +159,7 @@ def _validate_and_apply_step_config(self):
                 error_code=INVALID_PARAMETER_VALUE,
             )
 
+        self.output_probabilities = self.step_config.get("output_probabilities", False)
```
Can we change this `output_probabilities` to `predict_proba` to reduce the cognitive (over)load a bit?

prithvikannan committed:
rename to predicted_proba in logloss

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan committed:
rename to predict_proba

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan committed:
merging master

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

prithvikannan committed:
merge master

Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>
